heudiconv/utils.py

```diff
@@ -662,32 +664,107 @@ def get_datetime(date: str, time: str, *, microseconds: bool = True) -> str:
 
     Parameters
     ----------
     date_string: str
-      Date string to parse
-    fmt: str
-      Format string. If it ends with [.%f], we keep it if date_string ends with
-      '.\d+' regex and not if it does not.
+      Datetime string to parse
     """
+    # https://bids-specification.readthedocs.io/en/stable/common-principles.html#units
+    fmts = ["%Y-%m-%dT%H:%M:%S.%f%z", "%Y-%m-%dT%H:%M:%S%z", "%Y-%m-%dT%H:%M:%S.%f", "%Y-%m-%dT%H:%M:%S"]
+    datetime_obj = strptime(datetime_string, fmts)
+    return datetime_obj
 
-    optional_micr = "[.%f]"
-    if fmt.endswith(optional_micr):
-        fmt = fmt[: -len(optional_micr)]
-        if re.search(r"\.\d+$", date_string):
-            fmt += ".%f"
-    return datetime.strptime(date_string, fmt)
+def strptime_dcm_da_tm(dcm_data: dcm.Dataset, da_tag: TagType, tm_tag: TagType) -> datetime:
+    r"""
+    Create a datetime object from a dicom DA tag and TM tag.
 
+    Parameters
+    ----------
+    dcm_data : dcm.FileDataset
+        DICOM with header, e.g., as read by pydicom.dcmread.
+        Objects with __getitem__ and have those keys with values properly formatted may also work
+    da_tag: str
+      Dicom tag with DA value representation
+    tm_tag: str
+      Dicom tag with TM value representation
+    """
+    # https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_6.2.html
+    date_str = dcm_data[da_tag].value
+    fmts = ["%Y%m%d",]
+    date = strptime(date_str, fmts)
+
+    time_str = dcm_data[tm_tag].value
+    fmts = ["%H", "%H%M", "%H%M%S", "%H%M%S.%f"]
+    time = strptime(time_str, fmts)
+
+    datetime_obj = datetime.datetime.combine(date.date(), time.time())
+
+    if (0x0008, 0x0201) in dcm_data:
+        utc_offset = dcm_data[0x0008, 0x0201].value
+        datetime_obj = datetime_utc_offset(datetime_obj, utc_offset) if utc_offset else datetime_obj
+    return datetime_obj
+
+def strptime_dcm_dt(dcm_data: dcm.Dataset, dt_tag: TagType) -> datetime:
+    r"""
+    Create a datetime object from a dicom DT tag.
+
+    Parameters
+    ----------
+    dcm_data : dcm.FileDataset
+        DICOM with header, e.g., as read by pydicom.dcmread.
+        Objects with __getitem__ and have those keys with values properly formatted may also work
+    da_tag: str
+      Dicom tag with DT value representation
+    """
+    # https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_6.2.html
+    datetime_str = dcm_data.get(dt_tag)
+    fmts = ["%Y%z", "%Y%m%z", "%Y%m%d%z", "%Y%m%d%H%z", "%Y%m%d%H%M%z", "%Y%m%d%H%M%S%z", "%Y%m%d%H%M%S.%f%z",
+            "%Y", "%Y%m", "%Y%m%d", "%Y%m%d%H", "%Y%m%d%H%M", "%Y%m%d%H%M%S", "%Y%m%d%H%M%S.%f"]
+    datetime_obj = strptime(datetime_str, fmts)
+
+    if not datetime_obj.tzinfo and (0x0008, 0x0201) in dcm_data:
+        utc_offset = dcm_data[0x0008, 0x0201].value
+        datetime_obj = datetime_utc_offset(datetime_obj, utc_offset) if utc_offset else datetime_obj
```

Comment:
I am not sure who we can receive a datetime_obj with timezone from above, but I think it better be explicitly checked and warned about so we do not quietly overload it:

```suggestion
    if utc_offset_dcm := dcm_data.get([0x0008, 0x0201]):
        if utc_offset := utf_offset_dcm.value:
             datetime_obj2 = datetime_utc_offset(datetime_obj, utc_offset)
             if datetime_obj.tzinfo and datetime_obj2 != datetime_obj:
                 lgr.warning("Unexpectedly previously parsed datetime %s contains zoneinfo which is different from the one obtained from DICOMs UTFOffset field: %s", datetime_obj, datetime_obj2)
             else:
                 datetime_obj = datetime_obj2 
```

or alike?